tracing: use DelegatingCarrier to avoid serialization

tbg opened the pull request:
Introduced this carrier type upstream. It allows us to embed the Span
information directly into our protos.
Will run the benchmarks since now we always embed a Span in outgoing RPCs.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4902)

<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[kv/txn_coord_sender.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4902#-KCBgDgV0LJGinrQUTCY:-KCBgDgV0LJGinrQUTCZ:678933252) ([raw file](https://github.com/cockroachdb/cockroach/blob/49d31ed37154d6079f3977e25ed818cc597f405b/kv/txn_coord_sender.go#L26)):*
missing import alias as is done elsewhere

---

*[util/tracing/span.go, line 19 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4902#-KCBflXItnl1imqyKXax:-KCBflXItnl1imqyKXay:-630779778) ([raw file](https://github.com/cockroachdb/cockroach/blob/49d31ed37154d6079f3977e25ed818cc597f405b/util/tracing/span.go#L19)):*
can you add a static assertion? for convenient jump-to-definition, if nothing else.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4902#-:-KCBgAJeJHE62Z7pjoG-:1417088721)*
<!-- Sent from Reviewable.io -->

tbg committed:
tracing: use DelegatingCarrier to avoid serialization

Introduced this carrier type upstream. It allows us to embed the Span
information directly into our protos. Chose to make *Trace a pointer in order
to be able to distinguish between no trace and a trace which happens to have
zero metadata.
